third_party/nvfuser/csrc/scheduler/normalization_utils.cpp

```diff
@@ -494,5 +495,195 @@ std::optional<GridOuterNormalizationParams> getGridOuterNormalizationParams(
+    TensorView* tv0) {
+  std::unordered_set<TensorView*> all_producer_tvs;
+  std::queue<TensorView*> unvisited;
+
+  // start search from tv0, save its direct consumers
+  for (auto tv : ir_utils::producerTvsOf(tv0)) {
+    // skip reduction tv since it is allowed to have shared producer for
+    // reduction tvs.
+    if (!tv->hasReduction() &&
+        all_producer_tvs.find(tv) == all_producer_tvs.end()) {
+      all_producer_tvs.emplace(tv);
+      unvisited.push(tv);
+    }
+  }
+  // search for indirect consumers from tv0's direct consumers
+  while (!unvisited.empty()) {
+    auto next_tv = unvisited.front();
+    unvisited.pop();
+    for (auto tv : ir_utils::producerTvsOf(next_tv)) {
+      if (!tv->hasReduction() &&
+          all_producer_tvs.find(tv) == all_producer_tvs.end()) {
+        all_producer_tvs.emplace(tv);
+        unvisited.push(tv);
+      }
+    }
+  }
+  return all_producer_tvs;
+}
+
+bool hasSharedConsumerNonReductionProducer(
+    const std::vector<TensorView*>& inner_reduction_tvs,
+    const std::vector<TensorView*>& outer_reduction_tvs) {
+  // step-1, check if inner reduction and outer reduction tvs have shared
+  // consumer convert inner_reduction_tvs to unordered_set
+  std::unordered_set<Val*> inner_reduction_tvs_set;
+  for (auto tv : inner_reduction_tvs) {
+    inner_reduction_tvs_set.emplace(tv);
+  }
+
+  // get all consumers of the inner_reduction_tvs
+  // shared consumers between inner reduction tvs are allowed.
+  const std::unordered_set<Val*>& all_consumer_tvs_inner =
+      DependencyCheck::getAllDependentVals(inner_reduction_tvs_set);
+
+  // check if outer reduction tvs have any shared consumer with inner reduction
+  // tvs and other outer reduction tvs
+  std::unordered_set<TensorView*> all_consumer_tvs_outer;
+  for (const auto otv : outer_reduction_tvs) {
+    const auto& out_vals = DependencyCheck::getAllDependentVals({otv});
+    const auto& out_tvs = ir_utils::filterByType<TensorView>(out_vals);
+    for (const auto tv : out_tvs) {
+      // check shared consumer with inner reduction tvs
+      if (all_consumer_tvs_inner.find(tv) != all_consumer_tvs_inner.end()) {
+        return true;
+      }
+      // check shared consumer with other outer reduction tvs
+      if (all_consumer_tvs_outer.find(tv) != all_consumer_tvs_outer.end()) {
+        return true;
+      } else {
+        all_consumer_tvs_outer.emplace(tv);
+      }
+    }
+  }
+
+  // step-2, check if consumers of inner reduction and outer reduction tvs have
+  // shared non-reduction producer. get all producer of inner reduction's
+  // consumers
+  std::unordered_set<TensorView*> all_producer_tvs_inner;
+  for (auto val : all_consumer_tvs_inner) {
+    if (auto tv = dynamic_cast<TensorView*>(val)) {
+      auto my_all_producer_tvs = getAllNonReductionProducerTvsOf(tv);
+      all_producer_tvs_inner.merge(my_all_producer_tvs);
+    }
+  }
+
+  // check producer of outer reduction's consumers
```

Comment:
Let's say there're two outer reductions and there's a tensor that is used by both of the consumer groups of the two outer reductions. It doesn't seem this analysis can detect such a tensor. The propagation from one outer reduction to another reduction seems to happen if such a tensor exists, no? Shouldn't we treat all of each inner and outer reductions as separate groups no matter it's inner or outer?

Reply:
This can be detected by:
```
      // check shared consumer with other outer reduction tvs
      if (all_consumer_tvs_outer.find(tv) != all_consumer_tvs_outer.end()) {
        return true;
      } else {
        all_consumer_tvs_outer.emplace(tv);
      }
```

Reply:
added code to detect links through consumers' producers
added test case.

Reply:
Thanks. Yes, I think that case should be fine now.

Just realized there's another issue. We haven't considered multi-output expressions. For tensors, `WelfordOp` is, I think, the only multi-output expression, so I don't think this would matter for now, but that doesn't mean there's no other multi-output expressions in the future.

The reason we need to consider multi-output expressions is that both transform and parallelization propagators enforce siblings have the same transformations and parallelization. Multiple output tensors of the same expression are generated within the same loop nest, so they must be transformed exactly in the same way with the same parallelization.

Handling multi-output expressions doesn't seem to work with the current algorithm. We need to grab producers then also their siblings, and from them grab consumers, and then producers of the consumers, etc. One way may be like:

```
std::unordered_set tv_group;
std::deque tensors_to_visit;
tensors_to_visit.push_back(reduction_output);
while (!tensors_to_visit.empty()) {
  auto tv = tensors_to_visit.back();
  tensors_to_visit.pop();
  tv_group.insert(tv);
  for (auto producer: producers_of(tv)) {
     if (producer not included in tv_group) {
        tensors_to_visit.push_back(producer);
     }
   }
   for (auto consumer: consumers_of(tv)) {
     if (consumer not included in tv_group) {
        tensors_to_visit.push_back(consumer);
     }
   }
   for (auto sibling: siblings_of(tv)) {
     if (sibling not included in tv_group) {
        tensors_to_visit.push_back(sibling);
     }
   }
}
```

This should visit all tensors that are reachable from `reduction_output`. We then just need to check if there's no overlap.

Or, the above traversal is actually what the transform propagator does, so maybe we could also customize it to do just traversal without modifying anything, i.e., dry run. If we could gather which tensors are traversed, we should be able to know which tensors would be affected by the transform propagator.

Reply:
added in `isConnectedOnlyThroughReductionProducer`